File: src/main/java/com/characohol/characohol_api/service/RecommendationService.java
package com.characohol.characohol_api.service;

import com.characohol.characohol_api.domain.Recommendation;
import com.characohol.characohol_api.repository.RecommendationRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

@Service
@RequiredArgsConstructor
public class RecommendationService {

    private final RecommendationRepository recommendationRepository;

    public List<Recommendation> getRecommendations() {
        return recommendationRepository.findAll();
    }

    public Recommendation getByTag(String tag) {
        return recommendationRepository.findByTag(tag).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
    }

    public Recommendation create(Recommendation recommendation) {
        return recommendationRepository.save(recommendation);
    }

    public Recommendation update(Long id, Recommendation recommendation) {
        Recommendation target = recommendationRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
        target.setTag(recommendation.getTag());
        target.setCharacter(recommendation.getCharacter());
        target.setMessage(recommendation.getMessage());
        return recommendationRepository.save(target);
    }

    public void delete(Long id) {
        if (!recommendationRepository.existsById(id)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND);
        }

        recommendationRepository.deleteById(id);
    }
}

File: src/main/kotlin/com/characohol/characohol_api/service/RecommendationService.kt
package com.characohol.characohol_api.service

import com.characohol.characohol_api.domain.Recommendation
import com.characohol.characohol_api.repository.RecommendationRepository
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.web.server.ResponseStatusException
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Service
class RecommendationService(private val recommendationRepository: RecommendationRepository) {

    fun getRecommendations(): Flux<Recommendation> {
        return recommendationRepository.findAll()
    }

    fun getByTag(tag: String): Mono<Recommendation> {
        return recommendationRepository.findByTag(tag)
            .switchIfEmpty(Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND)))
    }

    fun create(recommendation: Recommendation): Mono<Recommendation> {
        return recommendationRepository.save(recommendation)
    }

    fun update(id: Long, recommendation: Recommendation): Mono<Recommendation> {
        return recommendationRepository.findById(id)
            .switchIfEmpty(Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND)))
            .flatMap { target ->
                recommendationRepository.save(
                    Recommendation(
                        id = target.id,
                        tag = recommendation.tag,
                        character = recommendation.character,
                        message = recommendation.message
                    )
                )
            }
    }

    fun delete(id: Long): Mono<Void> {
        return recommendationRepository.existsById(id)
            .flatMap { exists ->
                if (exists) {
                    recommendationRepository.deleteById(id)
                } else {
                    Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND))
                }
            }
    }
}